apps/bondy/test/bondy_wamp_protocol_SUITE.erl

```diff
@@ -0,0 +1,211 @@
+    {raw, binary, bert},
+    {raw, binary, erl},
+    {raw, binary, json},
+    {raw, binary, msgpack},
+    {ws, binary, bert_batched},
+    {ws, binary, bert},
+    {ws, binary, erl_batched},
+    {ws, binary, msgpack_batched},
+    {ws, binary, msgpack},
+    {ws, text, json_batched},
+    {ws, text, json}
+]).
+-define(WAMP2_ENCODINGS, [
+    ?WAMP2_JSON,
+    ?WAMP2_MSGPACK,
+    ?WAMP2_BERT,
+    ?WAMP2_ERL, % not supported
+    ?WAMP2_MSGPACK_BATCHED,
+    ?WAMP2_JSON_BATCHED,
+    ?WAMP2_BERT_BATCHED,
+    ?WAMP2_ERL_BATCHED
+]).
+
+validate_subprotocol(_Config) ->
+    lists:foreach(
+        fun(SubProtocol) ->
+            ValidateResult = bondy_wamp_protocol:validate_subprotocol(SubProtocol),
+            case lists:member(SubProtocol, ?SUPPORTED_SUB_PROTOCOLS) of
+                true ->
+                    ?assertEqual({ok, SubProtocol}, ValidateResult);
+                false ->
+                    ?assertEqual({error, invalid_subprotocol}, ValidateResult)
+            end
+        end,
+        [{P, F, E} || P <- ?PROTOCOLS, F <- ?FRAMES, E <- ?ENCODINGS]),
+
+    lists:foreach(
+        fun(SubProtocolBinary) ->
+            SubProtocol = wamp_subprotocol:from_binary(SubProtocolBinary),
+            ValidateResult = bondy_wamp_protocol:validate_subprotocol(SubProtocolBinary),
+            case lists:member(SubProtocol, ?SUPPORTED_SUB_PROTOCOLS) of
+                true ->
+                    ?assertEqual({ok, SubProtocol}, ValidateResult);
+                false ->
+                    ?assertEqual({error, invalid_subprotocol}, ValidateResult)
+            end
+        end,
+        ?WAMP2_ENCODINGS),
+    
+    Error = {error, whatever},
+    ?assertEqual(Error, bondy_wamp_protocol:validate_subprotocol(Error)),
+
+    ?assertEqual({error, invalid_subprotocol},
+                 bondy_wamp_protocol:validate_subprotocol(<<"wamp.2.not.supported">>)).
+
+%% -----------------------------------------------------------------------------
+%% bondy_wamp_protocol:init
+%% -----------------------------------------------------------------------------
+
+init_error(_Config) ->
+    UnsupportedSubProtocol = {ws, binary, erl},
+    Peer = {{127,0,0,1}, 7},
+    Options = #{},
+    Result = bondy_wamp_protocol:init(UnsupportedSubProtocol, Peer, Options),
+    ?assertEqual({error, invalid_subprotocol, undefined}, Result).
+
+init_ok(_Config) ->
+    SubProtocol = {raw, binary, erl},
+    Peer = {{127,0,0,1}, 7},
+    Options = #{},
+    {ok, State} = bondy_wamp_protocol:init(SubProtocol, Peer, Options),
+
+    Context = bondy_wamp_protocol:context(State),
+    SessionId = bondy_context:session_id(Context),
+    ?assertEqual(SessionId, bondy_wamp_protocol:session_id(State)),
+    ?assertEqual(Peer, bondy_wamp_protocol:peer(State)),
+    ?assertEqual(undefined, bondy_wamp_protocol:agent(State)),
+    ?assertEqual(undefined, bondy_wamp_protocol:realm_uri(State)),
+    ?assertError(function_clause, bondy_wamp_protocol:ref(State)).
```

Comment:
Here I don't think we should have a `function_clause`. It should rather return `undefined`.